### ISSUE TITLE
Plugin did not respond due to fatal error: concurrent map writes

### ISSUE THREAD

**Issue by timcosta** [phillbaker/terraform-provider-elasticsearch#186]
Hey all -

I've been seeing some crashes with this provider version 1.5.7 and Terraform 1.0.0 due to concurrent map writes. The Terraform output says this is indicative of a bug in the provider and to let y'all know, so here's the full trace of what's happening:

```bash
elasticsearch_opendistro_roles_mapping.read_mapper: Destroying... [id=my_role_read]
elasticsearch_opendistro_roles_mapping.mapper: Destroying... [id=my_role]
╷
│ Error: Plugin did not respond
│
│ The plugin encountered an error, and failed to respond to the plugin.(*GRPCProvider).ApplyResourceChange call. The plugin logs may contain more details.
╵
╷
│ Error: Plugin did not respond
│
│ The plugin encountered an error, and failed to respond to the plugin.(*GRPCProvider).ApplyResourceChange call. The plugin logs may contain more details.
╵
Releasing state lock. This may take a few moments...

Stack trace from the terraform-provider-elasticsearch_v1.5.7 plugin:

fatal error: concurrent map writes

goroutine 40 [running]:
runtime.throw(0x1fc8cd2, 0x15)
	runtime/panic.go:1117 +0x72 fp=0xc0000deb30 sp=0xc0000deb00 pc=0x1035e12
runtime.mapassign_faststr(0x1dfab80, 0xc00007c8d0, 0xc000212340, 0x35, 0x0)
	runtime/map_faststr.go:211 +0x3f1 fp=0xc0000deb98 sp=0xc0000deb30 pc=0x10142f1
github.com/phillbaker/terraform-provider-elasticsearch/es.awsSession(0xc00003e121, 0x9, 0xc0003397a0, 0xc00026b2e8)
	github.com/phillbaker/terraform-provider-elasticsearch/es/provider.go:416 +0x389 fp=0xc0000df288 sp=0xc0000deb98 pc=0x1c87409
github.com/phillbaker/terraform-provider-elasticsearch/es.awsHttpClient(0xc00003e121, 0x9, 0xc0003397a0, 0x1)
	github.com/phillbaker/terraform-provider-elasticsearch/es/provider.go:436 +0x46 fp=0xc0000df2f8 sp=0xc0000df288 pc=0x1c87646
github.com/phillbaker/terraform-provider-elasticsearch/es.getClient(0xc0003397a0, 0x2e, 0xc00026b5e8, 0xc000188050, 0x4a)
	github.com/phillbaker/terraform-provider-elasticsearch/es/provider.go:269 +0x378 fp=0xc0000df4f0 sp=0xc0000df2f8 pc=0x1c84dd8
github.com/phillbaker/terraform-provider-elasticsearch/es.resourceElasticsearchOpenDistroRolesMappingDelete(0xc00032c700, 0x1d3f120, 0xc0003397a0, 0x24, 0x2976320)
	github.com/phillbaker/terraform-provider-elasticsearch/es/resource_elasticsearch_opendistro_roles_mapping.go:120 +0x24a fp=0xc0000df7d0 sp=0xc0000df4f0 pc=0x1ca532a
github.com/hashicorp/terraform-plugin-sdk/helper/schema.(*Resource).Apply(0xc00051dc20, 0xc00013cbe0, 0xc000202700, 0x1d3f120, 0xc0003397a0, 0x1, 0xc0000a5320, 0x1ed8360)
	github.com/hashicorp/terraform-plugin-sdk@v1.12.0/helper/schema/resource.go:283 +0x4a6 fp=0xc0000df860 sp=0xc0000df7d0 pc=0x1884fc6
github.com/hashicorp/terraform-plugin-sdk/helper/schema.(*Provider).Apply(0xc000290300, 0xc00026ba18, 0xc00013cbe0, 0xc000202700, 0xc0000a53a8, 0x0, 0x0)
	github.com/hashicorp/terraform-plugin-sdk@v1.12.0/helper/schema/provider.go:294 +0x99 fp=0xc0000df8c0 sp=0xc0000df860 pc=0x1883259
github.com/hashicorp/terraform-plugin-sdk/internal/helper/plugin.(*GRPCProviderServer).ApplyResourceChange(0xc000286510, 0x21a8550, 0xc0002b4f60, 0xc00032c150, 0xc000286510, 0xc0002b4f60, 0xc000241b80)
	github.com/hashicorp/terraform-plugin-sdk@v1.12.0/internal/helper/plugin/grpc_provider.go:885 +0x8a5 fp=0xc0000dfb20 sp=0xc0000df8c0 pc=0x18a9085
github.com/hashicorp/terraform-plugin-sdk/internal/tfplugin5._Provider_ApplyResourceChange_Handler(0x1f28ca0, 0xc000286510, 0x21a8550, 0xc0002b4f60, 0xc000200300, 0x0, 0x21a8550, 0xc0002b4f60, 0xc0000561e0, 0xe6)
	github.com/hashicorp/terraform-plugin-sdk@v1.12.0/internal/tfplugin5/tfplugin5.pb.go:3305 +0x214 fp=0xc0000dfb90 sp=0xc0000dfb20 pc=0x17bb534
google.golang.org/grpc.(*Server).processUnaryRPC(0xc0004f4900, 0x21b3778, 0xc00009b200, 0xc0002c4000, 0xc00007c690, 0x2937640, 0x0, 0x0, 0x0)
	google.golang.org/grpc@v1.27.1/server.go:1024 +0x522 fp=0xc0000dfe48 sp=0xc0000dfb90 pc=0x144e502
google.golang.org/grpc.(*Server).handleStream(0xc0004f4900, 0x21b3778, 0xc00009b200, 0xc0002c4000, 0x0)
	google.golang.org/grpc@v1.27.1/server.go:1313 +0xd2c fp=0xc0000dff68 sp=0xc0000dfe48 pc=0x14525ac
google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc000136420, 0xc0004f4900, 0x21b3778, 0xc00009b200, 0xc0002c4000)
	google.golang.org/grpc@v1.27.1/server.go:722 +0xab fp=0xc0000dffb8 sp=0xc0000dff68 pc=0x14602ab
runtime.goexit()
	runtime/asm_amd64.s:1371 +0x1 fp=0xc0000dffc0 sp=0xc0000dffb8 pc=0x106c881
created by google.golang.org/grpc.(*Server).serveStreams.func1
	google.golang.org/grpc@v1.27.1/server.go:720 +0xa5

goroutine 1 [chan receive]:
github.com/hashicorp/go-plugin.Serve(0xc000423ef8)
	github.com/hashicorp/go-plugin@v1.0.1/server.go:354 +0xbb4
github.com/hashicorp/terraform-plugin-sdk/plugin.Serve(0xc00063ff68)
	github.com/hashicorp/terraform-plugin-sdk@v1.12.0/plugin/serve.go:65 +0x1c5
main.main()
	github.com/phillbaker/terraform-provider-elasticsearch/main.go:10 +0x48

goroutine 49 [select]:
go.opencensus.io/stats/view.(*worker).start(0xc000290100)
	go.opencensus.io@v0.22.5/stats/view/worker.go:276 +0xcd
created by go.opencensus.io/stats/view.init.0
	go.opencensus.io@v0.22.5/stats/view/worker.go:34 +0x68

goroutine 7 [select]:
github.com/hashicorp/go-plugin.(*gRPCBrokerServer).Recv(0xc00007c2a0, 0x0, 0x0, 0x0)
	github.com/hashicorp/go-plugin@v1.0.1/grpc_broker.go:121 +0x86
github.com/hashicorp/go-plugin.(*GRPCBroker).Run(0xc0002940a0)
	github.com/hashicorp/go-plugin@v1.0.1/grpc_broker.go:411 +0x7a
created by github.com/hashicorp/go-plugin.(*GRPCServer).Init
	github.com/hashicorp/go-plugin@v1.0.1/grpc_server.go:80 +0x33c

goroutine 36 [syscall]:
os/signal.signal_recv(0x0)
	runtime/sigqueue.go:165 +0x9d
os/signal.loop()
	os/signal/signal_unix.go:23 +0x25
created by os/signal.Notify.func1.1
	os/signal/signal.go:151 +0x45

goroutine 37 [chan receive]:
github.com/hashicorp/go-plugin.Serve.func2(0xc0000a8120, 0x21b7db0, 0xc000280ae0)
	github.com/hashicorp/go-plugin@v1.0.1/server.go:342 +0x94
created by github.com/hashicorp/go-plugin.Serve
	github.com/hashicorp/go-plugin@v1.0.1/server.go:339 +0xb2b

goroutine 38 [IO wait]:
internal/poll.runtime_pollWait(0x85fbfa0, 0x72, 0x0)
	runtime/netpoll.go:222 +0x55
internal/poll.(*pollDesc).wait(0xc000290398, 0x72, 0x0, 0x0, 0x1fb97f3)
	internal/poll/fd_poll_runtime.go:87 +0x45
internal/poll.(*pollDesc).waitRead(...)
	internal/poll/fd_poll_runtime.go:92
internal/poll.(*FD).Accept(0xc000290380, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	internal/poll/fd_unix.go:401 +0x212
net.(*netFD).accept(0xc000290380, 0xc000286040, 0x0, 0x0)
	net/fd_unix.go:172 +0x45
net.(*UnixListener).accept(0xc00007d110, 0xc000286040, 0x0, 0x0)
	net/unixsock_posix.go:162 +0x32
net.(*UnixListener).Accept(0xc00007d110, 0xc0001f6e60, 0x18, 0xc00009a900, 0x144bcc5)
	net/unixsock.go:260 +0x65
google.golang.org/grpc.(*Server).Serve(0xc0004f4900, 0x21a2cd8, 0xc000288b20, 0x0, 0x0)
	google.golang.org/grpc@v1.27.1/server.go:597 +0x27f
github.com/hashicorp/go-plugin.(*GRPCServer).Serve(0xc000290500, 0x21a2cd8, 0xc000288b20)
	github.com/hashicorp/go-plugin@v1.0.1/grpc_server.go:131 +0x6e
created by github.com/hashicorp/go-plugin.Serve
	github.com/hashicorp/go-plugin@v1.0.1/server.go:353 +0xb9a

goroutine 19 [select]:
google.golang.org/grpc/internal/transport.(*controlBuffer).get(0xc00013c230, 0x1, 0x0, 0x0, 0x0, 0x0)
	google.golang.org/grpc@v1.27.1/internal/transport/controlbuf.go:395 +0xff
google.golang.org/grpc/internal/transport.(*loopyWriter).run(0xc0001242a0, 0x0, 0x0)
	google.golang.org/grpc@v1.27.1/internal/transport/controlbuf.go:513 +0x1dd
google.golang.org/grpc/internal/transport.newHTTP2Server.func2(0xc00009b200)
	google.golang.org/grpc@v1.27.1/internal/transport/http2_server.go:296 +0xd7
created by google.golang.org/grpc/internal/transport.newHTTP2Server
	google.golang.org/grpc@v1.27.1/internal/transport/http2_server.go:293 +0x1110

Error: The terraform-provider-elasticsearch_v1.5.7 plugin crashed!

This is always indicative of a bug within the plugin. It would be immensely
helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.
```

I've seen this happen fairly frequently, and I think usually during deletes and state refreshes but honestly I can't be totally sure that I haven't seen it during writes either.

Generally re-running the apply 1-2 more times resolves the issue, so I'm fairly confident this isn't an issue with our cluster or my networking.

**Comment by phillbaker** [comment]
Sorry for the hassle, this is likely due to https://github.com/phillbaker/terraform-provider-elasticsearch/commit/52c17745997f8910556ed6f34249d05528b18f19#diff-d66ec846962229ad0d1815062f5a913f12a1f3eee175033bc7c3e8634d015a80, I think the map writes were not actually necessary to fix the problem, the config changes were enough. Let me verify that and see if I can revert the map usage.

**Comment by timcosta** [comment]
No worries at all, thank you for looking into this so quickly!

**Comment by phillbaker** [comment]
Should be fixed with b7add6d
